### PR TITLE
Clarify Read trait blocking behavior

### PR DESCRIPTION
**Pull request by ivmarkov**
This PR is an effort to clarify the blocking behavior of the `Read` trait with regards to the length of the user-supplied buffer.

It is a follow up on a recent discussion in the ["Rust Embedded" Matrix chat](https://matrix.to/#/!BHcierreUuwCMxVqOf:matrix.org/$KPz0XkHW0cQt2bAG2-8pGvzZNMSeZlTJlVTBlVVVEvY?via=matrix.org&via=catircservices.org&via=tchncs.de) where the agreement was that we could probably be more explicit about that.

For now, I've changed only the documentation of the blocking `Read` trait. Once/if we arrive at a text which seems reasonable to everybody, I'll extend the PR with a similar change to `embedded_io_async::Read`.

Ideally, we should arrive at a piece of documentation that we could also justify to be included upstream [in the Rust STD Read trait as well](https://github.com/rust-lang/rust/blob/edbc000fa4413f9ff92f4e5492f8ff497716ce48/library/std/src/io/mod.rs#L595), and possibly in its `async` variants in the `futures` and `tokio` crates.

### CODE REVIEW

**Comment by ivmarkov** [comment]
@Dirbaio @jamesmunns Pinging you here as you were actively participating in the Rust Embedded discussion on the subject.

File: embedded-io/src/lib.rs
Comment (jamesmunns) [inline embedded-io/src/lib.rs:305]:
```suggestion
    /// - Once at least one (or more) bytes become available, a non-zero amount of those is copied to the
```

**Review by jamesmunns — COMMENTED** [review]
Overall I'm find with this addition, I think it is very clearly written!

My only questions are whether we say "SHOULD" or "is recommended", to make clear that this isn't mandatory, at least until we have matching language in `std::io::Read`, or whether we note "this is how the `recv` syscall works in posix, and we're aiming to match that behavior which is generally consistent in all major desktop operating systems".

**Comment by ivmarkov** [comment]
> Overall I'm find with this addition, I think it is very clearly written!
> 
> My only questions are whether we say "SHOULD" or "is recommended", to make clear that this isn't mandatory, at least until we have matching language in `std::io::Read`, or whether we note "this is how the `recv` syscall works in posix, and we're aiming to match that behavior which is generally consistent in all major desktop operating systems".

I'm very certain it cannot be "SHOULD" or "recommended". It is a "MUST", so to say, or else any protocol implementation on top of `Read` that assumes this behavior is simply buggy and needs to be rewritten to read only byte-by-byte.

That includes [the "read_until" readers in STD](https://github.com/rust-lang/rust/blob/edbc000fa4413f9ff92f4e5492f8ff497716ce48/library/std/src/io/mod.rs#L2087) and the [`fill_buf`](https://github.com/rust-lang/rust/blob/edbc000fa4413f9ff92f4e5492f8ff497716ce48/library/std/src/io/buffered/bufreader/buffer.rs#L136) implementation.

File: embedded-io/src/lib.rs
Comment (jannic) [inline embedded-io/src/lib.rs:310]: Did you intentionally remove this paragraph?

While there may be no literal end-of-file for embedded-io, there are similar concepts like a closed pipe, where it doesn't make sense to continue blocking, and there's no data available to return.

Comment (ivmarkov) [reply embedded-io/src/lib.rs:310]: Nope - sorry  - this was completely unintentional. Re-added and thanks for spotting!

**Comment by MabezDev** [comment]
> I'm very certain it cannot be "SHOULD" or "recommended". It is a "MUST", so to say, or else any protocol implementation on top of Read that assumes this behavior is simply buggy and needs to be rewritten to read only byte-by-byte.

I would agree with this, otherwise we end up in the same place where one implementor may choose not to follow this "recommendation" and we end up with differing impls.

This LGTM from my perspective once we've applied the same to the async docs.

File: embedded-io/src/lib.rs
Comment (Dirbaio) [inline embedded-io/src/lib.rs:312]: minor nit: this line looks like it applies only to the "if there are bytes available to read" case, while it should apply to both. Perhaps split it to a new paragraph and remove the `-`.

**Comment by Dirbaio** [comment]
LGTM. Thanks for taking care of this :+1: 

(I too think it should be MUST, for the already-stated reasons)

File: embedded-io/src/lib.rs
Comment (ivmarkov) [reply embedded-io/src/lib.rs:312]: Hmmm, but it _is_ applicable only to the "if there are bytes available to read" case, as it talks about - if there are bytes available to read _in the first place_ - how many of those would be returned - i.e. not necessarily all of them (because (a) the buffer might actually be shorter than what is available (b) because the implementation just decides so, for whatever other reasons).

**Comment by ivmarkov** [comment]
I've now added identical changes to the `async` version of the `Read` trait, where the only changes I did is to use "waiting" at the places where the blocking version uses either "blocking" or "blocking and waiting".

With that said, shall we delay the merge a bit until I open a similar PR or issue for the `std::io::Read` trait? *If* they accept a similar change, they might have more feedback to the text, that we might want to incorporate in our `Read` trait versions too?

**Comment by Dirbaio** [comment]
Up to you. Getting feedback or merges to rust-lang PRs is not exactly fast though :)

**Comment by ivmarkov** [comment]
> Up to you. Getting feedback or merges to rust-lang PRs is not exactly fast though :)

I want to mitigate the danger of making the described behavior a MUST in our traits, and get resistance from the STD team to do it for the `Read` trait. 
- If they just return feedback that it is "obvious" - fine... 
- ...but if they say that it is unspecified i.e. "Option B" (what to do with the line-readers and stuff is then another topic), that would be quite a problem...

**Comment by Dirbaio** [comment]
IMO even if std explicitly decides to make it Option B we should go for Option A.

Option B is just insane. as you point out many things like ReadBuf / read_line become broken unless you read byte by byte.

**Comment by adamgreig** [comment]
We briefly discussed this in the meeting today as well, it seems like merging this now seems the best option. We can always wait a bit before releasing it to see what happens on rust-lang.

**Comment by ivmarkov** [comment]
> We briefly discussed this in the meeting today as well, it seems like merging this now seems the best option. We can always wait a bit before releasing it to see what happens on rust-lang.

OK - then be it. I've anyway not even started the PR/issue in STD upstream.

So if you could look at the `embedded_io_async::Read` trait wording as well, and if no additional feedback, I guess we can merge.

**Comment by ivmarkov** [comment]
If no further feedback, can somebody with committer permissions approve this?

File: embedded-io/src/lib.rs
Comment (jannic) [reply embedded-io/src/lib.rs:312]: This applies to the first case as well, doesn't it? In "Once at least one (or more) bytes become available", one could also ask if it's guaranteed that all available bytes are returned. The answer is "it is not guaranteed", but it's not completely obvious from the description as it is written now.

But as Dirbaio wrote, it is a "minor nit". Personally, I would not expect to have such a guarantee when reading the text you proposed.

**Review by jannic — APPROVED** [review]

File: embedded-io/src/lib.rs
Comment (ivmarkov) [reply embedded-io/src/lib.rs:312]: Not sure how come the first case is unclear that only a subset of the just-read bytes are returned, given that it says (emphasis mine): 
"
- Once at least one (or more) bytes become available, **a non-zero amount of those** is copied to the beginning of `buf`, and the amount is returned...

"

"a non-zero amount of those". Do we really need to be more explicit than that?

Comment (ivmarkov) [reply embedded-io/src/lib.rs:312]: Ah be it. If two folks find it unclear then it is probably unclear. :(
I'll re-arrange so that the problematic text is a separate paragraph.

Comment (ivmarkov) [reply embedded-io/src/lib.rs]: @jannic I've re-arranged it now so that the problematic text is a separate paragraph. If you can take a quick look - thanks.

Comment (jannic) [reply embedded-io/src/lib.rs]: > Not sure how come the first case is unclear that only a subset of the just-read bytes are returned, given that it says (emphasis mine): "
> 
>     * Once at least one (or more) bytes become available, **a non-zero amount of those** is copied to the beginning of `buf`, and the amount is returned...
> 
> 
> "
> 
> "a non-zero amount of those". Do we really need to be more explicit than that?

IMHO the only possible source of confusion was the asymmetry between the two cases. Yes, in general I agree that the wording "a non-zero amount of those" was sufficiently clear. But then the second case, in addition to that, explicitly stated that it's allowed to return less than available. For the first case, that sentence was missing. Readers may wonder why? Is there a difference in guarantees?

Again, this was a _very_ minor detail. But if we can avoid this possible confusion by just putting the last sentence in a separate paragraph, as you now did, we should do it.

**Review by jannic — APPROVED** [review]

File: embedded-io/src/lib.rs
Comment (ivmarkov) [reply embedded-io/src/lib.rs]: Argh, don't want to over-complicate it, but _now_ - with the problematic text being a separate paragraph - the _next_ paragraph that starts with "This waiting behavior is important for the cases where `Read`..." becomes unclear... in that readers might be confused that it only refers to the newly-born "Note..." paragraph, while it refers to _everything_ mentioned above.

Any brilliant ideas? Or is it just my mind that finds the new arrangement also problematic?

**Review by MabezDev — APPROVED** [review]
LGTM, thanks for taking care of this!

**Pull request merged**
